### ISSUE TITLE
Incorrect props Infer error when used Intersection types as template parameter in React.ComponentType

### ISSUE THREAD

**Issue by qqingvarqq** [facebook/flow#7946]
Flow version: v.0.103.0 or v.0.102.0 
[Try flow link](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVBLAtgBzgJwBdkwBDAZzACUBTUgY2KnzizAHJ87H2BudGgA88RMIQCeOGmCgZ85YgF4wAb1RgwAaxriAjAC4wC-BgB2AcwA0qAL79UQkcQlSjNenFMATMMrUbtcQAmQ2MzK1t7RwJnSWkAIzhCAAtfGTkFMAAyNw9vflAwVABIGgA3GnxxFPCwDEo4TTAEDBS4AFdiM0JK8ndCDE8AOgdhGLE4sESUtLViwIMwi2t5nRCliLtUFAEx0RdpD1xPGlMlam5CIYBhVjxTU8IAFTiAHmnUnJUk5MqAaR0hlIpnENgAfPYoO1TIxBqYwOYaIRbscHmcABQASkMR3uj1U6jAXEI7Xw8JwLBwlEUYLAry8GDKYJUFLgVKGP3+Ohsr2ADKZ-BsqDymRRODSiORdxOGMxUT2xC8NCgpHaMGIWN8YPRrzFWh0ukUACJdEb9cFjUEzZz8ADxMbTcAwXKgA)
## Actual behavior
```javascript
/* @flow */

import * as React from 'react';

export type first = {
  key1: string,
};

export type second = {
  key2: string,
};

export type both = first & second;
/* 
everything is ok without intersection.
export type both = {
	key1:string,
	key2:string,
};
*/

export type component = React.ComponentType<both & {otherKey: any}>;

function getComponent(): component {
  return props => <div>{props.otherKey}/* error here when both is intersection type*/</div>;
}
const Comp = getComponent();

export default () =>(<Comp key1="1" key2="2" otherKey="1"/>);
```
```
25:   return props => <div>{props.otherKey}</div>;
                                  ^ Cannot get `props.otherKey` because property `otherKey` is missing in props [1].
References:
22: export type component = React.ComponentType<both & {otherKey: any}>;
                                                ^ [1]
```
## Expected behavior
No error when intersection type used.

**Comment by taylorgoolsby** [comment]
I have a similar problem. This is what my code looks like:

```
import type { SessionStore } from '../../stores/SessionStore'
import type { LogicStore } from '../../stores/LogicStore'

type Stores = {
  sessionStore: SessionStore,
  logicStore: LogicStore,
}

type Props = {
  data: ApolloResponse,
}

class Who extends Component<Props & Stores> {
  renderHeader = () => {
    const { logicStore } = this.props
  }

  renderItem = () => {
    const { logicStore } = this.props
  }
}
```

Error is:
`Property logicStore is missing in Props [1].`

But this is strange. I only get the error for `renderItem`. The reference to `logicStore` in `renderHeader` is fine.

`SessionStore` and `LogicStore` are class definitions.
I.E. `export class SessionStore {}`

`ApolloResponse` is:
```
export type ApolloResponse = {
  viewer?: Viewer,
} & GraphqlQueryControls<any>
```

**Comment by taylorgoolsby** [comment]
Update:

I fixed the error by changing the offending `const { logicStore } = this.props` to `const logicStore = this.props.logicStore`. This error went away, but a new error popped up which wasn't there before:
`Cannot call logicStore.choose with item.id bound to itemId
because undefined [1] is incompatible with string [2].`

I fixed the signature of `logicStore.choose` to `choose(itemId: ?string)` and everything worked.
In fact, changing the code back to `const { logicStore } = this.props`, there were no errors.

So now I'm pretty sure this is a bug having to do with flow correctly recognizing there is a type error somewhere, but it's not reporting the error in the right place and it's not giving the correct error message.

In other words,
```
  renderItem = (item: Item) => {
    const { logicStore } = this.props
    return (
      <div onClick={() => logicStore.choose(item.id)}/> 
    )
  }
```
should give an error if `choose(itemId: string)`, and `type Item = {id: ?string}`.

However, flow is reporting it incorrectly because it is instead saying:
`Property logicStore is missing in Props [1].`

**Comment by taylorgoolsby** [comment]
It gets worse.

In addition to the problems I described above,
destructuring assignment, like `const { logicStore } = this.props`, also has a problem where if I later dereference a property which does not exist, like `logicStore.nonExistFoo`, then `logicStore` will be infered as `any (implicit)`.

For example, if I have:
```
const { logicStore } = this.props
logicStore.nonExistFoo()
```

This will not throw any type errors because `logicStore` is inferred to be `any`.
Doing a `type-at-pos` over `logicStore` on the line `logicStore.nonExistFoo()` returns `any (implicit)`.
